scripts/centralize_docs.py

```diff
@@ -0,0 +1,180 @@
+#! /usr/bin/env python3
+
+import pathlib
+import json
+import tempfile
+import subprocess
+import sys
+from collections import defaultdict
+from copy import deepcopy
+
+
+def _load_jsonl(file):
+    with open(file, 'r') as f:
+        _d = [json.loads(line) for line in f]
+        try:
+            data = {int(doc['metadata']['pmcid']): doc for doc in _d}
+        except:
+            print(f'Error loading {file}')
+            return None
+    return data
+
+
+def check_annotations(annotations, docs):
+    """ Given an annotation like follows:
+
+    {'end_byte': 20027,
+    'end_char': 19906,
+    'label_name': 'male',
+    'start_byte': 20023,
+    'start_char': 19902}
+
+    check if the annotation is still valid in the new document.
+    
+    """
+    for annotation in annotations:
+        old_doc = docs[0]
+        new_doc = docs[1]
+
+        old_text = old_doc['text']
+        new_text = new_doc['text']
+
+        old_start = annotation['start_byte']
+        old_end = annotation['end_byte']
+
+        old_text = old_text[old_start:old_end]
+
+        new_text = new_text[old_start:old_end]
+
+        if old_text != new_text:
+            return False
+
+    return True
+
+
+def get_new_documents(temp_dir_path, pmcids):
+
+    # Save pmids to a new file
+    pmcids_file = temp_dir_path / 'pmcids.txt'
+    pmcids = [str(pmcid) for pmcid in pmcids]
+    pmcids_file.write_text('\n'.join(pmcids))
+
+    # Run pubget run command using subprocess module
+    out_dir = temp_dir_path / 'output'
+    subprocess.run(['pubget', 'download', '--pmcids_file',
+                    str(pmcids_file), str(out_dir)])
+
+    res_dir = list(out_dir.glob('pmcidList*'))[0]
+
+    subprocess.run(['pubget', 'extract_articles', 
+                    str(res_dir / 'articlesets')])
+
+    subprocess.run(['pubget', 'extract_data', 
+                    str(res_dir / 'articles')])
+
+    subprocess.run(['pubget', 'extract_labelbuddy_data',
+                    str(res_dir / 'subset_allArticles_extractedData')])
+
+    json_files = list(
+        (res_dir / 'subset_allArticles_labelbuddyData').glob('*.jsonl'))
+
+    # Load old and new documents
+    new_docs = {}
+    for file in json_files:
+        new_docs.update(_load_jsonl(file))
+
+    return json_files, new_docs
+
+
+def get_old_documents():
+    old_docs = {}
+    all_pmids = set()
+    for file in pathlib.Path('projects').glob('*/documents/*.jsonl'):
+        # Skip symlinked files
+        if file.is_symlink():
+            continue
+        docs = _load_jsonl(file)
+
+        if not docs:
+            # Skip documents that don't have pmcids
+            continue
+
+        old_docs[file] = docs
+        all_pmids.update(docs.keys())
```

Comment:
I'm not sure if we can indentify documents by their pmcid, because we could have several versions of the same pmcid with slightly different text. that's in theory, I don't know if it's actually the case I'll try to check

Reply:
actually it seems there are 28 such documents, only 5 of which have annotations:

```
sqlite> select pmcid, count(distinct doc_md5) as n_md5 from detailed_annotation
   ...> where pmcid is not null
   ...> group by pmcid
   ...> having n_md5 > 1;
pmcid    n_md5
-------  -----
3888398  2    
3903437  2    
3913832  2    
3984441  2    
4183380  2    
4352055  2    
5082970  2    
5317331  2
```

Reply:
Yes this is true. What I did was check if the previous annotations would be compatible with a fresh download, and if so then I assume the differences between the two version is not relevant.